plugins/ccgx/fu-ccgx-hpi.c

```diff
@@ -0,0 +1,1633 @@
+		return FALSE;
+	}
+
+	row_size = ccgx_info->flash_row_size;
+	device_data->fw_row_size = row_size;
+
+	if (row_size > CYACD_FLASH_ROW_MAX) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "Not row size");
+		return FALSE;
+	}
+
+	row_max = ccgx_info->flash_size / row_size;
+	row_buffer = g_malloc0(row_size);
+
+	if (row_size == 128) {
+		fw_meta_offset = HPI_META_DATA_OFFSET_ROW_128;
+	}
+	else if (row_size == 256) {
+		fw_meta_offset = HPI_META_DATA_OFFSET_ROW_256;
+	} else {
+		fw_meta_offset = 0;
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "not support row size");
+		return FALSE;
+	}
+
+	device_data->fw_meta_offset = fw_meta_offset;
+	device_data->fw1_meta_row_num = row_max-1;
+	device_data->fw2_meta_row_num = row_max-2;
+
+	device_data->fw_meta_valid = FALSE;
+	if (!fu_ccgx_hpi_cmd_enter_flash_mode(device,hpi_handle,error)) {
+		return FALSE;
+	}
+
+	g_usleep (HPI_CMD_ENTER_FLASH_MODE_DELAY_US);
+
+	/* fw1 meta data */
+	fw_index = FW_MODE_FW1;
+	fw_meta_row_num = device_data->fw1_meta_row_num;
+
+	(void)fu_ccgx_hpi_clear_all_event (device, hpi_handle, 10, error);
+
+	if (!fu_ccgx_hpi_cmd_read_flash (device, hpi_handle, fw_meta_row_num, row_buffer, row_size,error)) {
+		return FALSE;
+	}
+
+	metadata = &device_data->fw_metadata[fw_index];
+	memcpy(metadata,&row_buffer[fw_meta_offset],sizeof(CCGxMetaData));
+
+	fw_version = &device_data->fw_version[fw_index];
+	if (metadata->metadata_valid == CCGX_METADATA_VALID_SIG) {
+		if (!fu_ccgx_hpi_get_fw_version(device,  hpi_handle, (metadata->last_boot_row + 1), row_size ,fw_version, error)) {
+			return FALSE;
+		}
+	}
+
+	/* fw2 meta data */
+	fw_index = FW_MODE_FW2;
+	fw_meta_row_num = device_data->fw2_meta_row_num;
+
+	if (!fu_ccgx_hpi_cmd_read_flash (device, hpi_handle, fw_meta_row_num, row_buffer, row_size, error)) {
+		return FALSE;
+	}
+
+	metadata = &device_data->fw_metadata[fw_index];
+	memcpy(metadata,&row_buffer[fw_meta_offset],sizeof(CCGxMetaData));
+
+	fw_version = &device_data->fw_version[fw_index];
+	if (metadata->metadata_valid == CCGX_METADATA_VALID_SIG) {
+		if (!fu_ccgx_hpi_get_fw_version(device,  hpi_handle, (metadata->last_boot_row + 1), row_size ,fw_version, error)) {
+			return FALSE;
+		}
+	}
+
+	if (!fu_ccgx_hpi_cmd_leave_flash_mode(device,hpi_handle,error)) {
+		return FALSE;
+	}
+
+	device_data->fw_meta_valid = TRUE;
+	g_usleep (HPI_CMD_ENTER_FLASH_MODE_DELAY_US);
+	return TRUE;
+}
+
+/**
+ * fu_ccgx_hpi_cmd_enter_flash_mode:
+ * @device: #FuDevice
+ * @hpi_handle HPI handle
+ * @error: a #GError or %NULL
+ *
+ * Enter flash mode in PD I2C Device
+ *
+ * Returns: %TRUE for success
+*/
+gboolean
+fu_ccgx_hpi_cmd_enter_flash_mode (FuDevice *device, CyHPIHandle *hpi_handle,  GError **error)
+{
+	CyPDResp hpi_event = 0;
+
+	g_return_val_if_fail (hpi_handle != NULL, FALSE);
+
+	(void)fu_ccgx_hpi_clear_all_event(device,hpi_handle,HPI_CMD_COMMAND_CLEAR_EVENT_TIME_MS,error);
```

Comment:
This seems wrong, don't you need to look at the return value and show an error?

Reply:
OK. I will make fu_ccgx_hpi_clear_all_events get process GError and get return value when it has error.